lib/parse.js

```diff
@@ -8,33 +8,73 @@ var defaults = {
     allowDots: false,
     allowPrototypes: false,
     arrayLimit: 20,
+    charset: 'utf-8',
     decoder: utils.decode,
     delimiter: '&',
     depth: 5,
+    interpretNumericEntities: false,
     parameterLimit: 1000,
     plainObjects: false,
-    strictNullHandling: false
+    strictNullHandling: false,
+    utf8Sentinel: false
 };
 
+var interpretNumericEntities = function (str) {
+    return str.replace(/&#(\d+);/g, function ($0, numberStr) {
+        return String.fromCharCode(parseInt(numberStr, 10));
+    });
+};
+
+// This is what browsers will submit when the ✓ character occurs in an
+// application/x-www-form-urlencoded body and the encoding of the page containing
+// the form is iso-8859-1, or when the submitted form has an accept-charset
+// attribute of iso-8859-1. Presumably also with other charsets that do not contain
+// the ✓ character, such as us-ascii.
+var isoSentinel = 'utf8=' + encodeURIComponent('&#10003;'); // %26%2310003%3B
+
+// These are the percent-encoded utf-8 octets representing a checkmark, indicating
+// that the request actually is utf-8 encoded.
+var utf8Sentinel = 'utf8=' + encodeURIComponent('✓'); // %E2%9C%93
+
 var parseValues = function parseQueryStringValues(str, options) {
     var obj = {};
     var cleanStr = options.ignoreQueryPrefix ? str.replace(/^\?/, '') : str;
     var limit = options.parameterLimit === Infinity ? undefined : options.parameterLimit;
     var parts = cleanStr.split(options.delimiter, limit);
+    var charset = options.charset;
+    var i;
+
+    if (options.utf8Sentinel) {
+        for (i = 0; i < parts.length; ++i) {
+            if (parts[i].indexOf('utf8=') === 0) {
+                if (parts[i] === utf8Sentinel) {
+                    charset = 'utf-8';
```

Comment:
That's exactly the intention -- that the character set inferred from the sentinel can should take precedence. When you specify both a `charset` and `utf8Sentinel:true`, the `charset` acts as more of a default charset.

*  [prefers an iso-8859-1 charset specified by the utf8 sentinel to a default charset of utf-8](https://github.com/papandreou/qs/blob/228956bab2ee33f59e1d0c5352ce4e7015a566e6/test/parse.js#L595-L598)
* [does not interpret numeric entities when the charset is utf-8, even when the interpretNumericEntities option is given](https://github.com/papandreou/qs/blob/228956bab2ee33f59e1d0c5352ce4e7015a566e6/test/parse.js#L630-L633)

Reply:
hmm. when an explicit charset is provided, it seems to me that the user is saying "i don't care what you think it is, treat it like it's this" - if it's a default charset, i'd expect it to be called `defaultCharset` when used along with the utf8 sentinel option. Thoughts?

Reply:
Hmm, yeah, I see the point, even though adding another option also contributes to the complexity.

Then `qs.parse(..., { utf8Sentinel: true, charset: 'iso-8859-1' })` would throw because that combination of options don't make sense?

Another concern is the symmetry with `qs.stringify`. Here `qs.stringify(..., { utf8Sentinel: true, charset: 'iso-8859-1'})` *does* make sense. It'd be a bit weird to force `defaultCharset` for `parse`, but not `stringify`.

On balance I think it'd be best to keep it as `charset` in all cases, but very explicitly point out how the `charset`+`utf8Sentinel` combo is interpreted in the `utf8Sentinel` documentation. WDYT?

Reply:
hmm. consistency between parse and stringify is important, for sure.

I think maybe parse should indeed throw on `{ utf8Sentinel: true, charset: 'anything that's not utf-8' }`, although that makes sense for stringify (where it should work) - and then we can keep the option called "charset".

Reply:
That would prohibit my original use case for these additions: A service compatible with the ancient `FormMail.pl` script (originally written in the nineties). It has to assume `iso-8859-1` because of all the ancient HTML snippets on iso-88591-1 pages pointed at it (with no sentinel).
However, more “modern” pages can explicitly switch to utf-8 mode by including the sentinel.
This requires the `utf8Sentinel: true, charset: ‘iso-8859-1` combo.

Reply:
ahhh that makes sense.

alrighty, thanks for bearing with me on the discussion. let's keep what you've got.